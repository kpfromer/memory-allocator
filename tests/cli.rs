use assert_cmd::prelude::*;
use std::process::Command;
use std::str;

type TestResult = Result<(), Box<dyn std::error::Error>>;

#[test]
fn prints_valid_lru() -> TestResult {
    let mut cmd = Command::cargo_bin("memory-replacement")?;
    cmd.args(&["1 2 3 4 1 2 5 1 2 3 4 5", "-f", "3", "-t", "lru"]);
    let out = cmd.assert().success();
    insta::assert_snapshot!(str::from_utf8(&out.get_output().stdout)?);

    Ok(())
}


#[test]
fn prints_valid_fifo() -> TestResult {

    // 3 frames
    let mut cmd = Command::cargo_bin("memory-replacement")?;
    cmd.args(&["1 2 3 4 1 2 5 1 2 3 4 5", "-f", "3", "-t", "fifo"]);
    let out = cmd.assert().success();
    insta::assert_snapshot!(str::from_utf8(&out.get_output().stdout)?);

    // 4 frames
    let mut cmd = Command::cargo_bin("memory-replacement")?;
    cmd.args(&["1 2 3 4 1 2 5 1 2 3 4 5", "-f", "4", "-t", "fifo"]);
    let out = cmd.assert().success();
    insta::assert_snapshot!(str::from_utf8(&out.get_output().stdout)?);
    
    Ok(())
}

#[test]
fn prints_valid_opt() -> TestResult {
    // 3 frames
    let mut cmd = Command::cargo_bin("memory-replacement")?;
    cmd.args(&["1 2 3 4 1 2 5 1 2 3 4 5", "-f", "3", "-t", "opt"]);
    let out = cmd.assert().success();
    insta::assert_snapshot!(str::from_utf8(&out.get_output().stdout)?);

    Ok(())
}

#[test]
fn prints_valid_opt_2() -> TestResult {
    // 4 frames
    let mut cmd = Command::cargo_bin("memory-replacement")?;
    cmd.args(&["1 2 3 4 1 2 5 1 2 3 4 5", "-f", "4", "-t", "opt"]);
    let out = cmd.assert().success();
    insta::assert_snapshot!(str::from_utf8(&out.get_output().stdout)?);

    Ok(())
}