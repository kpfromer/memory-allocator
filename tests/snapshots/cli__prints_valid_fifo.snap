---
source: tests/cli.rs
expression: "str::from_utf8(&out.get_output().stdout)?"

---
1 2 3 4 1 2 5 1 2 3 4 5
Hits: 3 - Misses: 9
+---+---+---+---+---+---+---+---+---+---+---+---+
| 1 | 1 | 1 | 4 | 4 | 4 | 5 | 5 | 5 | 5 | 5 | 5 |
+---+---+---+---+---+---+---+---+---+---+---+---+
|   | 2 | 2 | 2 | 1 | 1 | 1 | 1 | 1 | 3 | 3 | 3 |
+---+---+---+---+---+---+---+---+---+---+---+---+
|   |   | 3 | 3 | 3 | 2 | 2 | 2 | 2 | 2 | 4 | 4 |
+---+---+---+---+---+---+---+---+---+---+---+---+

